### ISSUE TITLE
[Bug]: Lemmy does not handle HTTP `HEAD` requests properly

### ISSUE THREAD

**Issue by Security-Chief-Odo** [LemmyNet/lemmy#3902]
### Requirements

- [X] Is this a bug report? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a single bug? Do not put multiple bugs in one issue.
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.

### Summary

Lemmy server _by way of the actix-web module_ fails to properly handle HTTP `HEAD` requests. Any HTTP request with an empty body _HEAD in this case_ causes an error which presents as `404` status code. 

[RFC 7231](https://httpwg.org/specs/rfc7231.html#HEAD) says:
> The `HEAD` method is identical to `GET` except that the server **MUST NOT** send a message body in the response (i.e., the response terminates at the end of the header section). The server **SHOULD** send the same header fields in response to a `HEAD` request as it would have sent if the request had been a `GET`

Lemmy response behavior goes against the HTTP RFC spec and will possibly lead to this and other issues as clients see the wrong response. According to [actix documentation](https://docs.rs/actix-web/latest/actix_web/http/struct.Method.html) they provide a method to handle a `HEAD` request. 

Please implement the proper method handling for `HEAD` requests. 

---
See also [this actix-web issue](https://github.com/actix/actix-web/issues/2702),

### Steps to Reproduce

1. Install `curl` command
2. Run `curl -I` against a lemmy instance, receive error 404
3. Run `curl` without the -I against the same url, receive a 200

_See technical details for example_

### Technical Details

Standard issue `GET` request, returns HTTP 200 (OK)
```
$ curl -fSso /dev/null -w "Status: %{http_code}\n" https://dubvee.org/nodeinfo/2.0.json
Status: 200
```

Standard issue `HEAD` request _-I parameter_, returns HTTP 404 (Not Found)
```
$ curl -IfSso /dev/null -w "Status: %{http_code}\n" https://dubvee.org/nodeinfo/2.0.json
Status: 404
curl: (22) The requested URL returned error: 404 
```

`HEAD` request masquerading as a `GET` request _-I,  -X GET parameters_, returns HTTP 200 (OK)
```
$ curl -IfSso /dev/null -w "Status: %{http_code}\n" -X GET https://dubvee.org/nodeinfo/2.0.json
Status: 200
```

### Version

0.18.4

**Comment by Freakazoid182** [comment]
I don't necessarily agree that the Lemmy API goes against the HTTP RFC spec. The `HEAD` requests are simply not implemented and thus 404 is the correct response. I don't believe there's any spec that requires you to implement `HEAD` routes for every `GET` route. That being said, there's probably little harm in adding this. Personally I would only add this feature when there's an actual use case, not only theoretical ones. Still I did do a quick proof of concept :)

An option could be to simply add `head` routes for each `get` route. I've tested this by simply adding `head()` routes that execute the get handlers. The response return body is then empty. Actix seems to do handle this correctly.

[A sample branch](https://github.com/LemmyNet/lemmy/compare/main...Freakazoid182:lemmy:feature/http-head-methods)

Question is whether it would be configured for every possible route. There are distinct routes for `api`, `apub`, `webfinger`, `feeds`, `pictrs` and `nodeinfo`.
